api/envoy/extensions/filters/http/api_key_auth/v3/api_key_auth.proto

```diff
@@ -28,27 +26,66 @@ option (xds.annotations.v3.file_status).work_in_progress = true;
+// If none of the fields are set, then the filter will reject the request.
+message ApiKeyAuth {
+  message Credential {
+    // The value of the unique API key.
+    string api_key = 1 [(validate.rules).string = {min_len: 1}];
+
+    // The unique id or identity that used to identify the client or consumer.
+    string client_id = 2 [(validate.rules).string = {min_len: 1}];
+  }
+
+  message Credentials {
+    repeated Credential entries = 1;
+  }
+
+  // Api credentials used to authenticate the clients.
+  Credentials credentials = 1 [(udpa.annotations.sensitive) = true];
 
   // The header name to fetch the key.
   // If multiple values are present in the given header, the filter rejects the request.
-  // Only one of authentication_header, authentication_query, or authentication_cookie should be set.
-  string authentication_header = 2;
+  string authentication_header = 2
+      [(validate.rules).string =
+           {max_len: 1024 well_known_regex: HTTP_HEADER_NAME strict: false ignore_empty: true}];
 
   // The query parameter name to fetch the key.
-  // Only one of authentication_header, authentication_query, or authentication_cookie should be set.
-  string authentication_query = 3;
+  string authentication_query = 3 [(validate.rules).string = {max_len: 1024}];
 
   // The cookie name to fetch the key.
-  // Only one of authentication_header, authentication_query, or authentication_cookie should be set.
-  string authentication_cookie = 4;
+  string authentication_cookie = 4
+      [(validate.rules).string =
+           {max_len: 1024 well_known_regex: HTTP_HEADER_NAME strict: false ignore_empty: true}];
+}
+
+// API key auth configuration of per route or per virtual host or per route configuration.
+message ApiKeyAuthPerScope {
+  // Route specific APIKeyAuth configuration. This is optional and could be used to override the
+  // filter level ApiKeyAuth configuration **partly**.
+  //
+  // For example, if route specific ``authentication_header`` is set, then the filter will fetch
+  // the key from route specific ``authentication_header``. But the filter level ``keys`` will
+  // still be used. If route specific non-empty ``keys`` is set, then the filter will use the
+  // route specific ``keys`` to authenticate the client but the filter level ``authentication_header``
+  // will still be used. If both ``authentication_header`` and ``keys`` are set, then the filter
+  // will use the route specific ``authentication_header`` and ``keys`` to authenticate the client.
+  ApiKeyAuth override_config = 1;
+
+  // A list of client ids that are allowed to access the routes of the scope. If the list is empty,
+  // then all authenticated clients are allowed. This provides very limited but simple authorization.
+  // If more complex authorization is required, then use the :ref:`HTTP RBAC filter
+  // <config_http_filters_rbac>` instead.
+  repeated string allowed_clients = 2;
```

Comment:
I initially think we could re-use the `ApiKeyAuth` as the per route/vhost filter configuration. But after a thinking from control planes' perspective, I change my mind.

Considering that the api keys may be changed or updated for a client id/consumer id, it would be better to share same credential list for all route. So, when credential updates occur, the control plane needn't to iterate all routes to fine the routes that need to update.